wasm_exec/wasm_exec.go

```diff
@@ -0,0 +1,364 @@
+func (j *jsWasm) nanotime1(ctx context.Context, mod api.Module) uint64 {
+	return uint64(getSysCtx(mod).Nanotime(ctx))
+}
+
+// _walltime converts the GOARCH=wasm stack to be compatible with
+// api.ValueType in order to call walltime.
+func (j *jsWasm) _walltime(ctx context.Context, mod api.Module, sp uint32) {
+	sec, nsec := j.walltime(ctx, mod)
+	requireWriteUint64Le(ctx, mod.Memory(), "sec", sp+8, uint64(sec))
+	requireWriteUint32Le(ctx, mod.Memory(), "nsec", sp+16, uint32(nsec))
+}
+
+// walltime implements runtime.walltime which supports time.Now.
+//
+// See https://github.com/golang/go/blob/4170084ad12c2e14dc0485d2a17a838e97fee8c7/src/runtime/sys_wasm.s#L188
+func (j *jsWasm) walltime(ctx context.Context, mod api.Module) (uint64 int64, uint32 int32) {
+	return getSysCtx(mod).Walltime(ctx)
+}
+
+// _scheduleTimeoutEvent converts the GOARCH=wasm stack to be compatible with
+// api.ValueType in order to call scheduleTimeoutEvent.
+func (j *jsWasm) _scheduleTimeoutEvent(ctx context.Context, mod api.Module, sp uint32) {
+	delayMs := requireReadUint64Le(ctx, mod.Memory(), "delay", sp+8)
+	id := j.scheduleTimeoutEvent(ctx, mod, delayMs)
+	requireWriteUint32Le(ctx, mod.Memory(), "id", sp+16, id)
+}
+
+// scheduleTimeoutEvent implements runtime.scheduleTimeoutEvent which supports
+// runtime.notetsleepg used by runtime.signal_recv.
+//
+// Unlike other most functions prefixed by "runtime.", this both launches a
+// goroutine and invokes code compiled into wasm "resume".
+//
+// See https://github.com/golang/go/blob/4170084ad12c2e14dc0485d2a17a838e97fee8c7/src/runtime/sys_wasm.s#L192
+func (j *jsWasm) scheduleTimeoutEvent(ctx context.Context, mod api.Module, delayMs uint64) uint32 {
+	delay := time.Duration(delayMs) * time.Millisecond
+
+	resume := mod.ExportedFunction("resume")
+
+	// Invoke resume as an anonymous function, to propagate the context.
+	callResume := func() {
+		if err := j.failIfClosed(mod); err != nil {
+			return
+		}
+		// While there's a possible error here, panicking won't help as it is
+		// on a different goroutine.
+		_, _ = resume.Call(ctx)
+	}
+
+	return j.scheduleEvent(delay, callResume)
+}
+
+// _clearTimeoutEvent converts the GOARCH=wasm stack to be compatible with
+// api.ValueType in order to call clearTimeoutEvent.
+func (j *jsWasm) _clearTimeoutEvent(ctx context.Context, mod api.Module, sp uint32) {
+	id := requireReadUint32Le(ctx, mod.Memory(), "id", sp+8)
+	j.clearTimeoutEvent(id)
+}
+
+// clearTimeoutEvent implements runtime.clearTimeoutEvent which supports
+// runtime.notetsleepg used by runtime.signal_recv.
+//
+// See https://github.com/golang/go/blob/4170084ad12c2e14dc0485d2a17a838e97fee8c7/src/runtime/sys_wasm.s#L196
+func (j *jsWasm) clearTimeoutEvent(id uint32) {
+	if t := j.removeEvent(id); t != nil {
+		if !t.Stop() {
+			<-t.C
+		}
+	}
+}
+
+// _getRandomData converts the GOARCH=wasm stack to be compatible with
```

Comment:
@prep fyi here are my notes about go's use of getRandomData, though none of this is tested (hence draft)